### PR TITLE
Cast timestamp array to string array with timezone

### LINKED ISSUE #2607: Casting timestamp array to string should not ignore timezone
**Describe the bug**

`cast` kernel supports casting timestamp array to string array. But timezone of the timestamp array is ignored now. So the casting output strings are incorrect.

**To Reproduce**

Create a timestamp array with timezone, cast it to string type.

**Expected behavior**

Casting a timestamp array to string array should consider timezone.

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #2607.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by viirya** [comment]
cc @sunchao

File: arrow/src/compute/kernels/temporal.rs [hunk 1]
```diff
@@ -28,33 +28,33 @@ use chrono::format::{parse, Parsed};
 use chrono::FixedOffset;
 
 macro_rules! extract_component_from_array {
-    ($array:ident, $builder:ident, $extract_fn:ident, $using:ident) => {
+    ($array:ident, $builder:ident, $extract_fn:ident, $using:ident, $convert:expr) => {
```
Comment (sunchao) [inline arrow/src/compute/kernels/temporal.rs:31]: why we need the `$convert`?

Comment (viirya) [reply arrow/src/compute/kernels/temporal.rs:31]: The result type of `hour` etc is Int32Array, but these APIs of `Timelike` (`hour`, `minute`, `second` etc) return u32. Previously in this macro, it is hardcoded as something like `dt.$extract_fn() as i32`. But for `to_string` API that is called for `cast` kernel, we don't need any conversion. In order to reuse this macro, I add this `$convert` to extract `as i32` part.

File: arrow/src/compute/kernels/mod.rs [hunk 2]
```diff
@@ -35,5 +35,6 @@ pub mod sort;
 pub mod substring;
 pub mod take;
 pub mod temporal;
+// pub(crate) use temporal::extract_component_from_array;
```
[the remark this answers is not in the document]
Comment (viirya) [reply arrow/src/compute/kernels/mod.rs:38]: Oops, forgot to remove. This was left during debugging.

File: arrow/src/compute/kernels/cast.rs [hunk 3]
```diff
@@ -1482,25 +1487,35 @@ where
...
-                    array.value_as_datetime(ix).map(|v| v.to_string())
-                }
-            })
-            .collect::<GenericStringArray<OffsetSize>>(),
-    ))
+    let mut builder = GenericStringBuilder::<OffsetSize>::new();
+
+    if let Some(tz) = tz {
+        let mut scratch = Parsed::new();
+        extract_component_from_array!(
```
[the remark this answers is not in the document]
Comment (viirya) [reply arrow/src/compute/kernels/cast.rs:1505]: sure.

**Review by sunchao — APPROVED** [review]
LGTM

**Pull request merged**